source/common/http/header_map_impl.h

```diff
@@ -212,27 +228,58 @@ class HeaderMapImpl : NonCopyable {
           if (pseudo_headers_end_ == entry.entry_) {
             pseudo_headers_end_++;
           }
+          if (!lazy_map_.empty()) {
+            auto& values_vec = lazy_map_[entry.key().getStringView()];
+            if (values_vec.size() == 1) {
+              lazy_map_.erase(entry.key().getStringView());
+            } else {
+              values_vec.erase(std::remove_if(values_vec.begin(), values_vec.end(),
+                                              [&](HeaderNode it) { return it == entry.entry_; }),
+                               values_vec.end());
```

Comment:
Can you add some comments here? I've read this a few times and I'm not completely clear what is going on inside here

Reply:
Simplified a bit by splitting the erase and the remove_if calls, and added some comments.